From 707b696ddc62001693efd50578a6586c5e8339e4 Mon Sep 17 00:00:00 2001
From: Katsuya Matsubara <matsu@igel.co.jp>
Date: Wed, 24 Oct 2012 17:24:44 +0900
Subject: [PATCH 21/39] Once reinitialize an instance and then retry
 OMX_Init() if it fails

Reinitializing an instance in case that OMX_Init() fails may
be effective especially for some stateful OMXIL libraries since it
resets the state.
---
 omx/gstomx_util.c |    6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/omx/gstomx_util.c b/omx/gstomx_util.c
index 85de5f4..1175391 100644
--- a/omx/gstomx_util.c
+++ b/omx/gstomx_util.c
@@ -173,6 +173,7 @@ static inline GOmxImp *
 request_imp (const gchar * name, gboolean disable_postproc)
 {
   GOmxImp *imp = NULL;
+  int retry = 1;
 
   g_mutex_lock (imp_mutex);
   imp = g_hash_table_lookup (implementations, name);
@@ -187,6 +188,7 @@ request_imp (const gchar * name, gboolean disable_postproc)
     return NULL;
 
   g_mutex_lock (imp->mutex);
+reinit:
   if (imp->client_count == 0) {
     OMX_ERRORTYPE (*r_config) (OMX_STRING path);
     OMX_ERRORTYPE omx_error;
@@ -203,6 +205,10 @@ request_imp (const gchar * name, gboolean disable_postproc)
 
     omx_error = imp->sym_table.init ();
     if (omx_error) {
+      if (retry-- > 0) {
+        imp->sym_table.deinit ();
+        goto reinit;
+      }
       g_mutex_unlock (imp->mutex);
       return NULL;
     }
-- 
1.7.9.5

